lib/bots/notifybot.ts

```diff
@@ -0,0 +1,918 @@
+		return this.retrieveFileFromHash(fileRequest).then((fileContents: string | void) => {
+			if (!fileContents) {
+				throw new Error(`Couldn't find the old hash for the Keyframe: ${owner}/${repo}:${oldHash}`);
+			}
+
+			oldFile = fileContents;
+			fileRequest.hash = newHash;
+
+			return this.retrieveFileFromHash(fileRequest);
+		}).then((fileContents: string | void) => {
+			if (!fileContents) {
+				throw new Error(`Couldn't find the old hash for the Keyframe: ${owner}/${repo}:${newHash}`);
+			}
+			newFile = fileContents;
+
+			// Both of these files are YAML. Decode the YAML and look specifically for
+			// the versions of components, noting the old version and new version for each:
+			//  * No old version, we must go through every version in a Changelog
+			//  * Old version, we go through every version from old + 1 to new (might just be one)
+			//  * No new version, we abort
+			//  * Should *either* the old (if existing) or new version be invalid (ie. do not exist
+			//    in the component's CHANGELOG.md), then the component is skipped and an error reported.
+			const oldComponents = yaml.safeLoad(oldFile).keyframe.components || {};
+			const newComponents = yaml.safeLoad(newFile).keyframe.components;
+
+			// Track every new version by trying to match a component in the new version
+			// with a component in the old.
+			const componentVersions: ComponentTracker[] = [];
+			_.each(newComponents, (value, key: string) => {
+				// Try and find the old version of the component.
+				const newVersion = value.version;
+				const oldVersion = _.get(oldComponents[key], 'version', null);
+				const componentRepo = value.repository;
+
+				// If we have a new version, and it doesn't match the old version and has a repo,
+				// then we want to do something with it.
+				if (newVersion && repo && (newVersion !== oldVersion)) {
+					componentVersions.push({
+						oldVersion,
+						name: key,
+						newVersion,
+						repo: componentRepo,
+						incrementing: true,
+						versions: undefined
+					});
+				} else if (newVersion === oldVersion) {
+					// We don't operate on static versions.
+					this.logger.log(LogLevel.INFO, `Component version static, nothing to do for ${repo}`);
+				}
+			});
+
+			// We now have a set of components. For each of them, we now go through the process
+			// of following the Changelogs to find fixed issues.
+			// Any thrown error here needs to abort the run, as it means fundamentally something is
+			// wrong with the released versions.
+			return Promise.map(componentVersions, (component) => this.getNewPRs(component));
+		})
+		.filter((component) => _.has(component, 'versions'))
+		.then((components: ComponentTracker[]) => {
+			// We've a list of components and the PRs for each version.
+			// So, for each version we now harvest the relevant PRs and start the check
+			// backwards for Front URLs.
+			return Promise.map(_.filter(components, (comp) => comp.versions), (component: ComponentTracker) => {
```

Comment:
Do you still need to filter here, when you already filter the components just above?